### ISSUE TITLE
Why does Plots `copy` input arguments

### ISSUE THREAD

**Issue by mkborregaard** [JuliaPlots/Plots.jl#1175]
Came up on Slack: https://julialang.slack.com/archives/C690QRAA3/p1507911096000451
Ref here https://github.com/JuliaPlots/Plots.jl/blob/master/src/series.jl#L72-L88

This may be wasteful, and can cause problems when passing in types that don't have `copy` defined. Solutions are to avoid the `copy` or replace with `deepcopy`.

**Comment by timholy** [comment]
I would recommend against `deepcopy` for reasons of performance.

Not quite sure I understand the comment about "the error cropping up," but if it's a question of applying a systematic transformation (to all elements equally) you can use `MappedArrays`, assuming that the inputs are `AbstractArrays`. (If in contrast you need to alter values idiosyncratically, then you'll need the `copy`.)

Right now `MappedArrays` doesn't support `mapped(f, A, B...)`, but that could change (I think I even have a patch to do that locally that I just need to clean up).

**Comment by mkborregaard** [comment]
I don't understand the comment either - this is by @tbreloff . But unfortunately at this point in the code you can't assume that the input is an `AbstractArray`.

**Comment by timholy** [comment]
I tried taking out the `copy`, and I didn't get any additional failures. Assuming the test suite is sufficiently comprehensive, perhaps that's the answer?

**Comment by mkborregaard** [comment]
I think it could be the answer. The test suite is not comprehensive enough that I'd trust it blindly for something like this, so we should probably write some tests specifically focused on the issues that could arise here.
Most uses of `copy` in Plots (outside the highlighted place` is where `plotattributes` are copied to new series. `plotattributes` also contains the input data. Perhaps we could define a local copy for `KW` (the type of `plotattributes`that only copied the fields that might get changed (things such as color) but just passed the input data by reference.

**Comment by tbreloff** [comment]
Hey guys. I reluctantly added the copy because of some corner cases that I
can’t quite remember. I expect it will be fine in most cases, until it
isn’t. :/

Agreed that there’s a good chance the tests don’t cover those corner cases,
whatever they were.

I say remove the copy and wait for issues to be filed!

**Comment by mkborregaard** [comment]
Yeah I also tested it out locally and I even think it's a feature, as it makes `a = randn(10); p= plot(a); a[2] = 3; p` update :-)

**Comment by mkborregaard** [comment]
Makes animations fail. Will revisit Wednesday

**Comment by mkborregaard** [comment]
Here's the reason this makes example 2 fail: Normally, if there are 2 series, the input args are copied to the series. So e.g. if you have `x = 1:3; y1 = 2:4; y2 = 3:5; plot(x, [y1 y2])` the `x` will be copied onto both series. Now, if you `push!` to the plot, you'll `push!` both x and y first to series 1, then to series 2, which means in this case the same number gets pushed twice onto x. This means that x and y will no longer be the same length.
